omero2pandas/io_tools.py

```diff
@@ -0,0 +1,149 @@
+    Python file-like interface
+
+    Only a single RawFileStore instance can exist per BlitzGateway,
+    loading multiple readers will change the target file for the FileStore.
+
+    For optimal performance this class should be wrapped in a BufferedReader,
+    use the get_annotation_reader convenience function for this.
+    """
+    def __init__(self, conn: BlitzGateway, file_id: int, reporter=None):
+        LOGGER.debug(f"Creating new  for file {file_id}")
+        self._prx = conn.createRawFileStore()
+        self._file_id = file_id
+        self.open()
+        if not (size := self._prx.size()):
+            raise omero.ClientError(f"Invalid size for OriginalFile {file_id}")
+        self._size = size
+        self._offset = 0
+        self._reporter = reporter
+        if self._reporter is not None:
+            reporter.reset(total=size)
+
+    def __enter__(self):
+        self.open()
+
+    def __exit__(self, **kwargs):
+        self.close()
+        if self._reporter is not None:
+            self._reporter.close()
+
+    def open(self):
+        current = unwrap(self._prx.getFileId())
+        if current != self._file_id:
+            LOGGER.debug(f"Setting IO file ID to {self._file_id}")
+            self._prx.setFileId(self._file_id, {"omero.group": "-1"})
+
+    def read(self, size=-1):
+        # Read `size` bytes from the target object. -1 = 'read all'
+        LOGGER.debug(f"Reading {size} bytes from file {self._file_id}")
+        # Ensure we're reading the correct object
+        if size == -1:
+            size = self._size - self._offset
+        size = min(size, self._size - self._offset)
+        if not size or self._offset >= self._size:
+            LOGGER.debug("Nothing to read")
+            return b""
+        self.open()
+        data = self._prx.read(self._offset, size)
+        self._offset += size
+        if self._reporter is not None:
+            self._reporter.update(size)
+        return data
+
+    def readinto(self, buffer):
+        # Read bytes into the provided object and return number read
+        if self._offset >= self._size:
+            return 0
+        data = self.read(len(buffer))
+        count = len(data)
+        buffer[:count] = data
+        return count
+
+    def seek(self, offset, whence=os.SEEK_SET):
+        # Seek within the target file
+        if whence == os.SEEK_SET:
+            self._offset = offset
+        elif whence == os.SEEK_CUR:
+            self._offset += offset
+        elif whence == os.SEEK_END:
+            self._offset = self._size - offset
+        else:
+            raise ValueError(f"Invalid whence value: {whence}")
+        if self._offset > self._size:
+            self._offset = self._size
+        elif self._offset < 0:
+            self._offset = 0
+
+    def close(self):
+        file_id = unwrap(self._prx.getFileId())
+        if file_id == self._file_id:
+            # We only close if the current file is the active one
```

Comment:
It's not clear to me how this check will ever fail.

Reply:
Thanks @kkoz 

Much of this is down to odd behaviour in BlitzGateway. As far as I can tell the user can actually only create a single `RawFileStore` object, subsequent calls modify the existing instance. Therefore if they were to call and access another file while this object exists any subsequent calls would be interacting with whichever file they accessed last instead of the file our reader was made to interact with.

To solve this I've made the reader defensively check that the target file hasn't changed by calling `self.open()` before any reads. In most scenarios the active file ID will still be correct and nothing happens, but we don't want to risk reading the contents of a totally irrelevant file.

Reply:
Hmm - so is the idea that if the client hands you a BlitzGateway and you attempt to use the `RawFileStore` associated with it, they may also be using the `RawFileStore` associated with it and you'll be clobbering each other's `setFileId` calls? I would say that if that happens (i.e. client calls `setFileId` after `omero2pandas does`) we should throw, not just overwrite again. Otherwise we'll screw up whatever the client was trying to do.